Njw dispatch refactor

mo-nathan opened the pull request:
@JoeCohen here's how I was thinking about handling the dispatch concept.  Note that I spotted a bug in how ApplicationController#index is currently defined.  I don't think we can use class attributes since there isn't a good way to initialize the subclass versions of these attributes.  This in effect means that you'd have to explicitly set the ones you don't want to provide values which kind of defeats the purpose of having them in the parent class.  After playing around with it, I figured out that you could get the desired behavior by defining a couple of class level methods.  There may be a better way to do this in Ruby, but this seems to behave correctly.

JoeCohen committed:
Prototype refactor of index dispatch to sub-action

- Merges dispatch dictionary data with list of params.
- Keeps related data together
- Simplifies `index` method
- Shortens file.

mo-nathan committed:
Simplify index dispatch for CollectionNumbersController

mo-nathan committed:
Fix class variable initialization

JoeCohen commented:
Nice & super-DRY. Thank you!
- There may be one difference, re `always_index: true`. I''ll think about this and get back to you.

JoeCohen commented on app/controllers/collection_numbers_controller.rb:
```diff
@@ -3,39 +3,13 @@
-  #   :observation_index,
-  #   :show
-  # ]
+
+  # rubocop:disable Rails/LexicallyScopedActionFilter
   before_action :pass_query_params, except: :index
   before_action :store_location, except: [:index, :destroy]
+  # rubocop:enable Rails/LexicallyScopedActionFilter
```
```suggestion
  before_action :pass_query_params, only: [
    :show, :new, :create, :edit, :update, :destroy
  ]
  before_action :store_location, only: [
    :show, :new, :create, :edit, :update
  ]
```

mo-nathan replied:
Is the main reason for this to avoid disabling the cop?  I guess that seems reasonable.  I don't feel strongly about this, but I like that the except list is shorter.

nimmolo commented:
I like the simplicity. 

Incidentally, if we go this route, maybe consider renaming some index params that, although locally unambiguous, are not consistent with analogous params in other controllers: i'm thinking of the observations index especially. For example, instead of `:user`, `:project` — `:by_user`, `:for_project`. It might make them easier to remember. But i don't feel strongly about it.

JoeCohen commented:
I've satisfied myself that this PR doesn't change the behavior of `#index`.

(That behavior is sometimes surprising. For instance:
- https://mushroomobserver.org/collection_numbers/1 shows that collection number, while
- https://mushroomobserver.org/collection_numbers?id=1 gets an index of all collection numbers.

I think `index` essentially ignores the `id` param. 
But that's outside the scope of this PR.)

JoeCohen commented:
I agree that -- in the long run -- it would be good to rename the param
keys and make them as uniform as possible across different controllers and
scopes.
But I personally don't want to tackle it this PR because:
(a) it's a lot of work that's independent of, and shouldn't delay, the core
refactor;
(b) raises an issue about breaking 3rd party urls.

It would be nice to have param keys be at least

One little thing: "by_user" and "by: user" are (probably) pronounced the
same and look similar. It's a tiny bit more mental work to separate them.
See Clean Code, ch. 2 ("Use Pronounceable Names").
Could we avoid this issue? Maybe "created_by_x" and "sorted_by: x"?


On Sat, Feb 4, 2023 at 10:12 AM andrew nimmo ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1329#pullrequestreview-1284100294>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFGZLCR7ABXE3EFTC53WV2LXPANCNFSM6AAAAAAUQ3SC5M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>

pellaea commented:
Sorry to stick my nose in but this is a point I see come up a lot.

While I love having consistency, there is an issue with requiring code to
reflect external representations.  In this case we want (if I understand
correctly based on insufficient perusal of the code, sorry) to make sure
the parameter in the URL be the same as the method name?  And have the same
naming pattern across multiple controllers.  The latter makes a little more
sense to me, but they aren't entirely decoupled.  I think config/routes.rb
is the place to draw a firm boundary between what the URLs are and what the
code is.  This business where refactoring code requires changing URLs and
therefore breaking all sorts of browser caches and external links doesn't
make sense to me.

Maybe these issues are not really a problem in this particular case.  I
just want to keep raising this point so that people keep it in mind.  It
definitely has happened in the past.

On Sat, Feb 4, 2023 at 2:56 PM Joseph D. Cohen ***@***.***>
wrote:

> I agree that -- in the long run -- it would be good to rename the param
> keys and make them as uniform as possible across different controllers and
> scopes.
> But I personally don't want to tackle it this PR because:
> (a) it's a lot of work that's independent of, and shouldn't delay, the core
> refactor;
> (b) raises an issue about breaking 3rd party urls.
>
> It would be nice to have param keys be at least
>
> One little thing: "by_user" and "by: user" are (probably) pronounced the
> same and look similar. It's a tiny bit more mental work to separate them.
> See Clean Code, ch. 2 ("Use Pronounceable Names").
> Could we avoid this issue? Maybe "created_by_x" and "sorted_by: x"?
>
>
> On Sat, Feb 4, 2023 at 10:12 AM andrew nimmo ***@***.***>
> wrote:
>
> > ***@***.**** approved this pull request.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/MushroomObserver/mushroom-observer/pull/1329#pullrequestreview-1284100294
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAALDFGZLCR7ABXE3EFTC53WV2LXPANCNFSM6AAAAAAUQ3SC5M
> >
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***
> > com>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1329#issuecomment-1416837135>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYTNNK3BEHM4A3ECY5BNS3WV2X75ANCNFSM6AAAAAAUQ3SC5M>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

JoeCohen commented:
I was partly wrong. `id` is doing something. In particular:
https://mushroomobserver.org/collection_numbers?id=1 gets **the page containing collection number 1** of the index of all collection numbers.

JoeCohen commented:
Jason:
Thank you for chiming in and for the reminder about cache breakage.
I really didn't want to change the url's in this PR.
Now I really, really don't want to do it.
- Joe

On Sat, Feb 4, 2023 at 12:31 PM Jason Hollinger ***@***.***>
wrote:

> Sorry to stick my nose in but this is a point I see come up a lot.
>
> While I love having consistency, there is an issue with requiring code to
> reflect external representations. In this case we want (if I understand
> correctly based on insufficient perusal of the code, sorry) to make sure
> the parameter in the URL be the same as the method name? And have the same
> naming pattern across multiple controllers. The latter makes a little more
> sense to me, but they aren't entirely decoupled. I think config/routes.rb
> is the place to draw a firm boundary between what the URLs are and what the
> code is. This business where refactoring code requires changing URLs and
> therefore breaking all sorts of browser caches and external links doesn't
> make sense to me.
>
> Maybe these issues are not really a problem in this particular case. I
> just want to keep raising this point so that people keep it in mind. It
> definitely has happened in the past.
>
> On Sat, Feb 4, 2023 at 2:56 PM Joseph D. Cohen ***@***.***>
> wrote:
>
> > I agree that -- in the long run -- it would be good to rename the param
> > keys and make them as uniform as possible across different controllers
> and
> > scopes.
> > But I personally don't want to tackle it this PR because:
> > (a) it's a lot of work that's independent of, and shouldn't delay, the
> core
> > refactor;
> > (b) raises an issue about breaking 3rd party urls.
> >
> > It would be nice to have param keys be at least
> >
> > One little thing: "by_user" and "by: user" are (probably) pronounced the
> > same and look similar. It's a tiny bit more mental work to separate them.
> > See Clean Code, ch. 2 ("Use Pronounceable Names").
> > Could we avoid this issue? Maybe "created_by_x" and "sorted_by: x"?
> >
> >
> > On Sat, Feb 4, 2023 at 10:12 AM andrew nimmo ***@***.***>
> > wrote:
> >
> > > ***@***.**** approved this pull request.
> > >
> > > —
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/MushroomObserver/mushroom-observer/pull/1329#pullrequestreview-1284100294
> > >,
> > > or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AAALDFGZLCR7ABXE3EFTC53WV2LXPANCNFSM6AAAAAAUQ3SC5M
> > >
> > > .
> > > You are receiving this because you were mentioned.Message ID:
> > > ***@***.***
> > > com>
> > >
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/MushroomObserver/mushroom-observer/pull/1329#issuecomment-1416837135
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAYTNNK3BEHM4A3ECY5BNS3WV2X75ANCNFSM6AAAAAAUQ3SC5M
> >
> > .
> > You are receiving this because you are subscribed to this thread.Message
> > ID: ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1329#issuecomment-1416843865>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFF3NG2RLSXUQ6S5G2LWV24B3ANCNFSM6AAAAAAUQ3SC5M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nimmolo commented:
Agree this is not the PR for that... but LOL - you guys do realize... I just finished changing nearly every URL and parameter on the site! 

Especially in the case of indexes. Those are all routed through a single index route and not differentiated by routes anymore, so changing the name of an index param doesn't affect the routes. It will break links that have the param, but should not break cached queries. The only code that has to change is the param name, the params in the template links or forms that send or process that param, and the tests.

Of course the params themselves are debatable, but I've spent untold hours for no purpose other than to make this code more predictable and less taxing to learn for outsiders. So if a few param tweaks here and there would help that, breaking somebody's bookmarked but uncached link once again hardly seems like a reason not to do it.

nimmolo commented:
Also agree that i params don't need to strictly reflect action names.

But i think maybe that proposal was aiming to simplify the dispatch code... no strong feelings about it. 

The big advantage of consistency from a newbie point of view (I still feel like a newbie) is  that it makes the codebase much less mentally taxing to get a handle on. I can dip into a controller file and feel like i know where i'm going, when we repeat similar patterns.

JoeCohen committed:
Cover line added earlier in this PR

Fixes Coveralls check failure

JoeCohen commented:
Us beginners (beginner?) appreciate those changes. 👍
I don't know where I come out on breaking 3rd party urls.

On Sat, Feb 4, 2023 at 3:12 PM andrew nimmo ***@***.***>
wrote:

> Also agree that i params don't need to strictly reflect action names.
>
> But i think maybe that proposal was aiming to simplify the dispatch
> code... no strong feelings about it.
>
> The big advantage of consistency from a newbie point of view (I still feel
> like a newbie) is that it makes the codebase much less mentally taxing to
> get a handle on. I can dip into a controller file and feel like i know
> where i'm going, when we repeat similar patterns.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1329#issuecomment-1416873071>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFGQ2RB7FGHCUVBG7WDWV3O4DANCNFSM6AAAAAAUQ3SC5M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JoeCohen commented:
Yes that's the reason; the cop might help us out in the future.
(I also like shorter lists, but thought that the cop was useful. I also
don't feel strongly about it.)

On Sat, Feb 4, 2023 at 4:22 PM Nathan Wilson ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In app/controllers/collection_numbers_controller.rb
> <https://github.com/MushroomObserver/mushroom-observer/pull/1329#discussion_r1096602856>
> :
>
> > +
> +  # rubocop:disable Rails/LexicallyScopedActionFilter
>    before_action :pass_query_params, except: :index
>    before_action :store_location, except: [:index, :destroy]
> +  # rubocop:enable Rails/LexicallyScopedActionFilter
>
> Is the main reason for this to avoid disabling the cop? I guess that seems
> reasonable. I don't feel strongly about this, but I like that the except
> list is shorter.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1329#discussion_r1096602856>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFAA2WTAMN2GNZVJKRDWV3XFFANCNFSM6AAAAAAUQ3SC5M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>

mo-nathan commented:
> I agree that -- in the long run -- it would be good to rename the param keys and make them as uniform as possible across different controllers and scopes. But I personally don't want to tackle it this PR because: (a) it's a lot of work that's independent of, and shouldn't delay, the core refactor; (b) raises an issue about breaking 3rd party urls. It would be nice to have param keys be at least One little thing: "by_user" and "by: user" are (probably) pronounced the same and look similar. It's a tiny bit more mental work to separate them. See Clean Code, ch. 2 ("Use Pronounceable Names"). Could we avoid this issue? Maybe "created_by_x" and "sorted_by: x"?
> […](#)
> On Sat, Feb 4, 2023 at 10:12 AM andrew nimmo ***@***.***> wrote: ***@***.**** approved this pull request. — Reply to this email directly, view it on GitHub <[#1329 (review)](https://github.com/MushroomObserver/mushroom-observer/pull/1329#pullrequestreview-1284100294)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAALDFGZLCR7ABXE3EFTC53WV2LXPANCNFSM6AAAAAAUQ3SC5M> . You are receiving this because you were mentioned.Message ID: ***@***.*** com>

I like "created_by_x" and "sorted_by: x" (I was also thinking "order_by", but don't have a strong opinion).  "by" is kind of ambiguous.  However, I also agree that this change is out of scope for this PR.

mo-nathan commented:
I'm not terribly concerned about browser caches.  I'd rather not break external links, but it's hard to know how important that really is.  The current PR should break anything and the general mechanism allows us to continue to have backwards compatibility if we want.  At the same time I totally understand Nimmo's perspective on this.  Realistically less mental burden for new devs seems more important than maintaining external links as long as there isn't a huge amount of churn.  Ideally we'd put in something that mapped the old URLs to the new ones, but that's a pain and creates more clutter.

mo-nathan committed:
Update app/controllers/collection_numbers_controller.rb

Avoid disabling the cop.

Co-authored-by: Joseph D. Cohen <JoeCohen@users.noreply.github.com>

pellaea commented:
I feel like my point is being missed here -- no doubt because I'm presenting my case poorly.  I'm talking about very basic separation of concerns here.  If you develop a library or class that's used all over the place, you don't go around changing the public interface willy nilly because you want the public interface to more closely reflect internal refactoring.  Obviously.  Everyone here would presumably agree with this basic tenet of computer science!  And I'm not saying that's what's happening here, because I simply don't have time to read and understand what's going on.  But I see the symptoms: I see refactoring of code to cure ABC violations -- obviously purely internal -- and I see corresponding changes to the public interface -- URLs.  So I just wanted to raise this issue to make sure it's being considered.

mo-nathan commented:
I do see your concern and in general agree with the principle that such things should be adjustable independently if there is good reason for it.  I believe the code in this PR does that.  In my view the question is more about what do we want the external interface to be.  Obviously there is some weight for maintaining the existing interface.  However, if that interface is overly complex, inconsistent, or confusing, it may make sense to change even at the cost of breaking how things worked in the past.  What I think this exercise is telling us is that the existing interface could be cleaner.  Whether that cleanup is worth the cost is what we're discussing.  The AbcSize violations have already been fixed.  What those changes have revealed is that it might make sense to consider using better names in the params to avoid having to explicitly specify the mapping.  This follows in the Rails tradition of "intelligent defaults".  The trade off is whether changing the interface is justified by making the code simpler to comprehend by future and current developers.

pellaea commented:
Okay, that's fine, I feel heard.  As Nimmo said, that's what his CRUD refactoring has been all about for going on a year now! :)  And maybe this is yet another example of how Trailblazer's way of organizing concerns will make a common recurring problem (our URLs keep changing!) much less of an issue in the future.

JoeCohen committed:
Refactor CommentsController

per CollectionControllersNumber pattern

JoeCohen committed:
Remove obsolete comments

JoeCohen committed:
Disable LexicallyScopedActionFilter

Fixes LexicallyScopedActionFilter offense

JoeCohen committed:
Refactor LocationsDescriptionController#index

JoeCohen committed:
Refactor NamesDescriptionsController#index

JoeCohen committed:
Use a valid sort criterion to test by: x

fixes broken test.

JoeCohen committed:
Refactor HerbariumRecordsController#index

JoeCohen committed:
Tweak ImagesControllerTest

For uniformity with other controller test and for brevity

JoeCohen committed:
Refactor IMagesController#index

JoeCohen committed:
Refactor LocationsController#index

Includes fixing a test.

JoeCohen committed:
Refactor NamesControllerTest#index

JoeCohen committed:
Refactor ObservationsController#index

JoeCohen committed:
Delete extra space

JoeCohen committed:
Refactor ProjectsController#index

JoeCohen committed:
Conform SpeciesListsControllerTest method names

JoeCohen committed:
Test  SpeciesListController#index default sort order

JoeCohen committed:
Fix test expectation

JoeCohen committed:
Initial refactor of SpeciesListController#index

JoeCohen committed:
Add more tests for SpeciesListController

JoeCohen committed:
Simplify conditional

Fixes Style/SoleNestedConditional offense

JoeCohen committed:
Fix CommentsController index_subaction_param_keys

Deletes stray "by"
Fixes `NoMethodError: undefined method `by'`

JoeCohen commented:
🚧 In order to reduce email noise, I'm temporarily closing PR #1329 while I work on it. (I hope to re-open it by this evening.)

JoeCohen committed:
Fix HerbariumRecordsController sub-action names

Fixes test failures

JoeCohen committed:
Fix default sort order for ProjectsController

JoeCohen committed:
Merge branch 'main' into njw-dispatch-refactor

JoeCohen committed:
Fix Names pattern sub-action name

Fixes some failures created by the parent commit.

JoeCohen committed:
Fix LocationsController#index subaction

Fixes integration test Fails and Errors

JoeCohen committed:
Normalize LocationDescription Test index names

JoeCohen committed:
Augment test of LocationsDescription authored_by

- Checks if fixture still OK
- Tests visible page title
- Tests page contents

JoeCohen committed:
Test LocationsDescription#index with id param

JoeCohen committed:
Fix LocationsDescriptions#index for :by, :id, :q

JoeCohen committed:
Delete puzlling test that I skiped

JoeCohen committed:
uniform sub-action names for LocationsDescriptions#index

JoeCohen committed:
Get LocationsDescriptions default sort from Query

Instead of hard-coding it

JoeCohen committed:
Fix NamesDescriptionsControllerTest method names

Conforms them to LocationsDescriptionsControllerTest

JoeCohen committed:
update NamesDescriptionController sort order tests

conform to LocationDescriptionController

JoeCohen committed:
Test NamesDescriptionController#index id=n

JoeCohen committed:
Copy tests of LocationDescription#index

JoeCohen committed:
Merge branch 'main' into njw-dispatch-refactor

JoeCohen committed:
Merge branch 'main' into njw-dispatch-refactor

JoeCohen committed:
Merge branch 'main' into njw-dispatch-refactor

JoeCohen committed:
Fix test errors caused by prior merge of main

JoeCohen committed:
Standard Locations Descriptions test method names

JoeCohen committed:
Delete one instance of twice-defined method

JoeCohen committed:
Test LocationsDescriptions by_author with no descs

JoeCohen committed:
Add comment explaining reason for test

JoeCohen committed:
Test LocationsDescriptions by_author  bad user_id

JoeCohen committed:
Fix LocationsDescriptions by_author bad user_id

JoeCohen committed:
Document issue in LocationDescriptions fixture

JoeCohen committed:
Fix LocationDescriptions#index by_editor

- Adds tests
- `by_editor` subaction gets editor id from `:by_editor`, not `:id`

JoeCohen committed:
test NameDescription#index by_author of  1 desc

JoeCohen committed:
Fix test NameDescriptions by_aiuthor of >1 desc

JoeCohen committed:
Test NamesCriptions author of 0 descriptions

JoeCohen committed:
Fix testing NameDescriptons author of 1 desc

Directly accesses join table.

JoeCohen committed:
Test NameDescriptions by_editor of 1 desc

JoeCohen committed:
Test NameDescriptions by_editor > 1 desc

JoeCohen committed:
Test NameDescription by_editor of 0 descs

JoeCohen committed:
Tweak bad_user_id's

JoeCohen committed:
Delete draft and redundant tests

JoeCohen committed:
Define NameDescriptions default sort order

JoeCohen committed:
Conform NamesDescriptoinsController method names

JoeCohen committed:
Merge branch 'main' into njw-dispatch-refactor

JoeCohen committed:
Tweak NamesDescriptions list_all

- Mirrors current functionality and LocationsDescriptions
- Restores 100% coverage of #index

JoeCohen committed:
Fix CollectionNumbersControllerTest method names

- All names start with `test_index_`
- Other misc. changes to reveal test intention

JoeCohen committed:
Migrate and update schema

JoeCohen committed:
Fix CollectionNumbersController#index method names

- Conform to other controllers

JoeCohen committed:
Improve CollectionNumbersControllerTest

- Adds assertions re page content
- Deletes redundant methods
- Re-orders test methods

JoeCohen committed:
apply main by, q, id, logic to CollectionNumbers

JoeCohen committed:
Conform CommentsControllerTest mathod names

JoeCohen committed:
Conform CommentsController sub-action names

JoeCohen committed:
Test CommentsController#index target

JoeCohen committed:
Reorder & conform names CommentsController#index tests

JoeCohen committed:
Combine redundant tests

JoeCohen committed:
Re-order & -name Comments test index_by_user

JoeCohen committed:
CommentsController logic conformed to main

- If `:by` key exists, adds :by to current query , doesn't use :all flavor
- Define and use `default_sort_order`

JoeCohen committed:
Augment tests of CommentsController#index by_user

JoeCohen committed:
Augment tests of Comments for_user

- Adds tests for users who received 1 comment, 0 comments

JoeCohen committed:
Fix DoubleRender error for Comments by_user

JoeCohen committed:
Fix DoubleRender in Comments for_user

JoeCohen committed:
Comment re arrangement of CommentsController tests

JoeCohen committed:
Remove unneeded"@response" from call to redirect_url

- It's unnecessart

JoeCohen committed:
Reduce CommentsController#target Abc to <=17

JoeCohen committed:
Conform HerbariumRecordsController test names

JoeCohen committed:
Rearrange HerbariumRecordController tests

JoeCohen committed:
Tweak arrangement of CommentsController tests

JoeCohen committed:
Tweak HerbariumRecordsController test names

JoeCohen committed:
HerbariumRecord.count, instead of .all.size

JoeCohen committed:
Improve asserts of existing HerbariumRecord tests

- Specific page instead of just redirect
- Details of page contents
- Etc.

JoeCohen committed:
Disambiguate test method name

JoeCohen committed:
Conform HerbariumRecordsController aubaction names

JoeCohen committed:
Keep HerbariumRecordController logic re by, id, q

JoeCohen committed:
Define default_sort_order

JoeCohen committed:
Fix TrailingCommaInArrayLiteral offense

JoeCohen committed:
Rearrange ImageController#index tests

Uses standard method: names, order, format

JoeCohen committed:
Extract check_index_sorted_by

DRYs tests

JoeCohen committed:
Refactor test_index_pattern_image_id

Uses helper path in assertion

JoeCohen committed:
Improve ImageController advaced_search tests

Covers different resulting pages

JoeCohen committed:
Improve Images pattern_serach test

JoeCohen committed:
Standardize ImagesController#index method names

JoeCohen committed:
Fix ImagesController#index logic re :by

Uses same logic as main branch.

JoeCohen committed:
Define ImagesController#default_order

JoeCohen committed:
Fix ImageController#index Rubocop offenses

JoeCohen committed:
Delete extra blank line

JoeCohen committed:
Rearrange LocationsController#index tests

JoeCohen committed:
More detailed assertions re Locations country=

JoeCohen committed:
Merge branch 'main' into njw-dispatch-refactor

JoeCohen committed:
Revise Location#index?country= tests

- Uses one assertion per method
- Uses `assert_select` instead of `assert_template`.
See https://github.com/rails/rails/issues/18950#issuecomment-77924771

JoeCohen committed:
Accidentally omitted schema update

JoeCohen committed:
Add fixtures for testing Locations edited_by

- Adds new Location fixtures
- DRYs locations.yml with defaults
- Fixes bug in `unused_location` fixture
- Adds LocationsVersions fixtures for the added Location fixtures
- DRYs locations_versions.yml with defaults

JoeCohen committed:
Fix mistakes infixtures from prior commit

JoeCohen committed:
Enhance tests of Locations by_user

- Tests for page content
- More general test format

JoeCohen committed:
Fill out tests of Locations by_editor

Covers all logical case

JoeCohen committed:
Fill out tests of Locations by_user

JoeCohen committed:
Standardize Locations#index subactions

- Same order as dispatch table
- Standardize method names

JoeCohen committed:
Echo LocationsController#indexlogic re by, q, id

JoeCohen committed:
Update LocationsController documentation

JoeCohen committed:
Delete comma after last array element

JoeCohen committed:
Rubocop Locations stuff

Both Controller and ControllerTest

JoeCohen committed:
Delete redundant NamesControllerTest method

JoeCohen committed:
Rearrage NamesController#index tests

Standardize method names, method order, format

JoeCohen committed:
Use assert_select NamesController index tests

instead of assert_template

JoeCohen committed:
Fix Failure in Continuous Integration #6579

JoeCohen committed:
Add messages for assert_select("#title" ...)

JoeCohen committed:
Test NamesController#index advanced_search >1 hit

JoeCohen committed:
Delete unused constant

I added it earlier, realized I wouldn't use it, but forgot to delete it.

JoeCohen committed:
Test NamesController advanced_search 1 hit

JoeCohen committed:
Test NamesController pattern_search >1 hit

JoeCohen committed:
Rubocop fix spacing issues

JoeCohen committed:
Move part of controller test to integration test

Also DRYs remaining part of controller test

JoeCohen committed:
Augment Names test with_observations

Add assertions about page content

JoeCohen committed:
Augment test Names with_descriptions

Test right # of results

JoeCohen committed:
Augment Names need_descsriptions test

JoeCohen committed:
Change test name to  better show intent

JoeCohen committed:
Augment Names test by_user

JoeCohen committed:
Augment Names by_editor tests

JoeCohen committed:
Standardize Names#index sub-acetions

Uniform names, orders

JoeCohen committed:
Write test exposing bug in dispatching q to default method

JoeCohen committed:
Fix bug exposed by prior test

JoeCohen committed:
Define Names#default_sort_order

JoeCohen committed:
Move misplaced comment to right place

JoeCohen committed:
Rubocop Names index tests and controller

JoeCohen committed:
Standardize Observationsindex test names

JoeCohen committed:
Rearrange Observations index tests

Standardizes order and format.

JoeCohen committed:
Distribute Observations test_page_loads

Move gets and assertions (as needed) to separate methods

JoeCohen committed:
Augment Observations#index?id test

JoeCohen committed:
Split Observation Advanced Search notes tests

One request/methd.

JoeCohen committed:
Delete obsolete comment

JoeCohen committed:
Method wrapping assert_select(#title) in assertion

This gives more informative error messages.

JoeCohen committed:
Revert "Method wrapping assert_select(#title) in assertion"

This reverts commit 1a9a29155cbd4bcd7e48a1ea5aa4185edb39e74a.

JoeCohen committed:
General extenstion for assert_select(#title

- Wraps `assert_select` in an `assert`
- Gives better error msgs
- DRYer
- Reads better

JoeCohen committed:
Improve assert_title_id

- I had expect and go reversed
- Make it work with regexp expects
- And use it with regexps

JoeCohen committed:
More revisions to Names  pattern search tst

JoeCohen committed:
Change a Names pattern test to an integration test

JoeCohen committed:
Augment Names look_alikes tests

JoeCohen committed:
Delete extra argument in assert_title_id calls

Fixes failures in [CI test run 4310595942](/actions/runs/4310595942)

JoeCohen committed:
Augment Names test of related_taxa

JoeCohen committed:
Update Names#inder user tests

JoeCohen committed:
Trivial change to form of test_index_where

JoeCohen committed:
Update Observations#index :where param tests

Includes moving tests that really aren't of:where param

JoeCohen committed:
Add Observation#index tests re not on 1st page

JoeCohen committed:
Update Observations :projects tests

More precise assertions

JoeCohen committed:
Update test_index_with_lichen_filter

JoeCohen committed:
Fix typo in previous commit

Fixes test failuire

JoeCohen committed:
Standardize Names#index defaul subaction name

JoeCohen committed:
Rearrannge ObservationsIndex

- Privatizes everything
- Moves called methods to after last caller
- Standardizes `index` and `list` method names

JoeCohen committed:
Revise tests to expose bug in dispatch logic

JoeCohen committed:
Fix bug in dispatch logic

JoeCohen committed:
Get Observations default sort order from Query

JoeCohen committed:
Test  Observations#index with :needs_id

Covers one more line from ObservationsControllerTest
so that test covers Observations#index 100%

JoeCohen committed:
Rubocop ObservationsControllerTest

JoeCohen committed:
Fix ObservationsIndex non-metric offenses

JoeCohen committed:
Fix 2 Abc offenses in ObservationsIndex

JoeCohen committed:
Fix Observations advanced_search metric offenses

JoeCohen committed:
Standardize Projects index test method names

JoeCohen committed:
Merge branch 'main' into njw-dispatch-refactor

JoeCohen committed:
Augment Projects patternSearch tests

JoeCohen committed:
Standardize ProjectsController#index method names

JoeCohen committed:
Derive Projects default sort order from Query

JoeCohen committed:
Fix Project logic re :by param

JoeCohen committed:
Standardize SpeciesList index tests names, order

JoeCohen committed:
SpeciesLists index tests, more specific asserts

JoeCohen committed:
More tests of SepciesController for_project

JoeCohen committed:
Fix bug in SpeciesLists by_user

If selected user has not lists, it should display empty list
rather than list of SpeciesLists of current user

JoeCohen committed:
Additional test of SpeciesLists ffor_user

JoeCohen committed:
Delete redundant assertion

JoeCohen committed:
Get SpeciesLists default sort order from Query

JoeCohen committed:
Rubocop SpeciesListController

JoeCohen committed:
Restore controller dispatch tables

This is a cleaner solution then adding a conditional
to the default action to deal with presence of :by, :id, or :q
See discussion starting at
https://mushroomobserver.slack.com/archives/C040TH9FV/p1678029001202919

JoeCohen committed:
Rename list_query_results to index_query_results

More like the names in main which dsitinguish list... from index...

JoeCohen committed:
Rename assert_title_id to assert_displayed_title

`assert_title_id` is misleading
It suggests that we're checking the `id` of a `title`

JoeCohen committed:
Fix misuse of assert_flash_error

JoeCohen committed:
Replace template assertions with title assertions

JoeCohen committed:
Use assert_displayed_title vs assert_select

JoeCohen committed:
Restore.rubocop.yml inheritance

I had temporarily commented this out, and
accidentally pushed it.

JoeCohen committed:
Rubocopping

JoeCohen commented on app/controllers/comments_controller.rb:
```diff
@@ -53,51 +35,54 @@ class CommentsController < ApplicationController
-  def show_comments_by_user
-    user = find_or_goto_index(User, params[:by_user].to_s) || @user
+  def by_user
+    user = find_obj_or_goto_index(
+      model: User, obj_id: params[:by_user].to_s,
+      index_path: comments_path
+    )
```
This is a change in behavior:
List comments, instead of the current user's comments, If params[:by_user] is not a real user.

JoeCohen commented on app/controllers/comments_controller.rb:
```diff
@@ -106,24 +91,24 @@ def show_comments_for_user
-                                                  type: target.class.name)
-      show_selected_comments(query)
-    end
+  def target
```
Refactored to reduce Abc size to <= 17.

JoeCohen committed:
Derive HerbariumRecords default sort from Query

JoeCohen commented on app/controllers/images_controller.rb:
```diff
@@ -47,52 +39,54 @@ class ImagesController < ApplicationController
-      )
-      return
-    end
+  def list_all
```
Refactored to reduce Abc size to <= 17

JoeCohen committed:
heredoc instead of multi-line continued quote

JoeCohen committed:
Delete obsolete method alias

JoeCohen committed:
Fix comment

JoeCohen commented on app/controllers/observations_controller/index.rb:
```diff
@@ -1,223 +1,232 @@
-    query = create_query(:Observation, :at_location, location: location)
-    show_selected_observations(query)
-  end
+    def parents(name_str)
```
Does method belong in Name?
(There is a "Name#parents", but it's based on `classification`.)

JoeCohen committed:
Improve method name

JoeCohen committed:
Delete accidentally commited temporary files

JoeCohen committed:
Restore accidentally delted spacer

JoeCohen committed:
Narrow selector

Marginal improves efficiency

JoeCohen committed:
Update test method names

Start them with `test_index_` for easier bulk testing

JoeCohen committed:
Narrow more selectors

JoeCohen committed:
Improve some error messages

JoeCohen committed:
Narrower selector

JoeCohen committed:
Improve Observations test assertions

JoeCohen commented:
I'm finally done fiddling with this PR, so it's ready to go.
I plan on waiting until Fri. 10 March for any comments, and merging if there are none.
I'm not expecting review from @mo-nathan, as there are 34 files changed -- way over the limit of 20 that he mentioned. (I ended up modifying the tests as well as the controllers.)
